### PR TITLE
Improve performance of `Estimator` and `Sampler` by avoiding deep copy at `circuit_to_instruction`.

### PR DESCRIPTION
**Pull request by t-imamichi**
### Summary

This PR Improves performance of `Estimator` by converting circuits to instructions manually to avoid parameter binding in `circuit_to_instruction` method.

Address the same issue with #8367 

TODO
- [x] This PR might be over optimization. I need to check whether operations in `ciurcuit_to_instruction` can be optimized out or not.

### Details and comments

Comparison with microbenchmark
```python
from timeit import default_timer

from qiskit.circuit import QuantumCircuit
from qiskit.primitives import Estimator
from qiskit.quantum_info.random import random_pauli

num_qubits = 6
num_operators = 100
num_circuits = 100

circ = QuantumCircuit(num_qubits)
circ.x(range(num_qubits))
observables = [random_pauli(num_qubits, seed=seed) for seed in range(1, num_operators + 1)]

start = default_timer()
estimator = Estimator()
for _ in range(num_circuits):
    a = estimator.run([circ] * num_operators, observables)
print(f"{default_timer() - start} sec")
```

```
main branch
5.699802217999999 sec
```

```
this PR
3.949981157 sec
```

```
#8605 
5.321556266 sec
```

### CODE REVIEW

File: qiskit/primitives/estimator.py [hunk 1]
```diff
@@ -67,6 +68,17 @@ def __init__(
...
+    @staticmethod
+    def _bound_circuit_to_instruction(circuit: QuantumCircuit) -> Instruction:
+        inst = Instruction(
+            name=circuit.name,
+            num_qubits=circuit.num_qubits,
+            num_clbits=circuit.num_clbits,
+            params=[],
+        )
+        inst.definition = circuit
+        return inst
```
Comment (Cryoris) [inline qiskit/primitives/estimator.py:80]: Is this different from
```python
from qiskit.converters import circuit_to_instruction

inst = circuit_to_instruction(circuit)
```
? 🙂

**Comment by Cryoris** [comment]
Couldn't we update `circuit_to_instruction` to avoid the parameter binding if no parameters are present and speed up the function in general instead of just in the `Estimator`? 🙂

**Comment by t-imamichi** [comment]
Yes, it would be great if we can optimize `circuit_to_instruction` in a similar way.
We know that the output instruction won't be changed by Estimator. So, the optimization of this PR may work.
But, the output instruction of `circuit_to_instruction` could be manipulated in general. So, we might need to take care of it and the speed-up could be smaller.

**Comment by Cryoris** [comment]
In `circuit_to_instruction` L86 and following, couldn't we just do something like
```
 out_instruction = Instruction(
        name=circuit.name,
        num_qubits=circuit.num_qubits,
        num_clbits=circuit.num_clbits,
        params=[*parameter_dict.values()],
        label=label,
    )
    out_instruction.condition = None

-  target = circuit.assign_parameters(parameter_dict, inplace=False)
+  if circuit.num_parameters > 0:
+      target = circuit.assign_parameters(parameter_dict, inplace=False)
+  else:
+      target = circuit
```
that would avoid copying the circuit if we don't need to. Or is there also other overhead we would need to avoid?

**Comment by t-imamichi** [comment]
I worry that the output instruction might be perhaps affected if users modify the input circuit after calling `circuit_to_instruction`. Let me think of some cases.

**Comment by t-imamichi** [comment]
I notice a case where the output will be affected by the change of the input circuit. It may be difficult to optimize `circuit_to_instruction`.

```python
from qiskit.circuit import QuantumCircuit, Parameter
from qiskit.converters import circuit_to_instruction

qc = QuantumCircuit(1)
x = Parameter('x')
qc.rz(1, 0)
inst = circuit_to_instruction(qc)

print('before')
print(inst)
print(inst.definition)
print()

print('after')
qc[0].operation.params = [x]
print(inst)
print(inst.definition)
```

main branch
```
before
Instruction(name='circuit-78', num_qubits=1, num_clbits=0, params=[])
   ┌───────┐
q: ┤ Rz(1) ├
   └───────┘

after
Instruction(name='circuit-78', num_qubits=1, num_clbits=0, params=[])
   ┌───────┐
q: ┤ Rz(1) ├
   └───────┘
```

Julien's patch https://github.com/Qiskit/qiskit-terra/pull/8403?notification_referrer_id=NT_kwDOAdvAsLM0MDc4NjE4NTA0OjMxMTc4OTI4#issuecomment-1196848975
```
before
Instruction(name='circuit-78', num_qubits=1, num_clbits=0, params=[])
   ┌───────┐
q: ┤ Rz(1) ├
   └───────┘

after
Instruction(name='circuit-78', num_qubits=1, num_clbits=0, params=[])
   ┌───────┐
q: ┤ Rz(x) ├
   └───────┘
```
The definition of the output instruction changes when I modified the input circuit.

**Review by ikkoham — COMMENTED** [review]
This direction looks good to me. But I want to wait your TODO and Julien's comment.

**Comment by Cryoris** [comment]
Thanks a lot for checking @t-imamichi! Indeed it seems we cannot do this optimization on `circuit_to_instruction` since the instruction is added by reference. Would you mind adding a comment to the estimator code that explains why we only optimize that there? 

In this case LGTM!

**Comment by Cryoris** [comment]
This is also related to #7983 🙂

**Comment by t-imamichi** [comment]
Thanks. It's a good news. If `circuit_to_instruction` is updated, we don't need this PR anymore.

**Comment by t-imamichi** [comment]
I update the code to avoid this optimization when there are multiple qregs, which are supposed to be flattened by `circuit_to_instruction`. I think this code is ready.

**Review by ikkoham — APPROVED** [review]
Thanks. LGTM!

**Pull request merged**